src/tuple_impl.rs

```diff
@@ -280,23 +280,14 @@ macro_rules! impl_tuple_collect {
                 return None;
             }
 
-            #[allow(unused_assignments)]
             fn collect_from_iter_no_buf<I>(iter: I) -> Option<Self>
                 where I: IntoIterator<Item = $A>
             {
                 let mut iter = iter.into_iter();
-                loop {
-                    $(
-                        let $Y = if let Some($Y) = iter.next() {
-                            $Y
-                        } else {
-                            break;
-                        };
-                    )*
-                    return Some(($($Y),*,))
-                }
 
-                return None;
+                Some(($(
+                    { let $Y = iter.next()?; $Y },
+                )*))
```

Comment:
I *think* this modification is semantically equivalent, but I'd love if someone else could check it over before I merge.

Reply:
Looks good to me. Godbolt even seems to generate very similar code in `--release`.

off-topic: I do not have much time to spare at the moment, so I won't be very responsive.